micro_sam/models/peft_sam.py

```diff
@@ -56,6 +56,49 @@ def forward(self, x):
         return qkv
 
 
+class FacTSurgery(nn.Module):
+    """Operates on the attention layers for performing factorized attention.
+
+    (Inspired from: https://github.com/cchen-cc/MA-SAM/blob/main/MA-SAM/sam_fact_tt_image_encoder.py)
+
+    """
+
+    def __init__(
+            self,
+            rank: int,
+            block: nn.Module,
+            FacTu: nn.Module,
+            FacTv: nn.Module,
+    ):
+        super().__init__()
+        self.qkv_proj = block.attn.qkv
+        self.dim = self.qkv_proj.in_features 
+
+        self.q_FacTs = nn.Linear(rank, rank, bias=False)
+        self.v_FacTs = nn.Linear(rank, rank, bias=False)
+
+        # NOTE : Dropout is not included in the original implementation
+        self.dp_q = nn.Dropout(0.1)
+        self.dp_v = nn.Dropout(0.1)
```

Comment:
> Let's discuss whether or not it makes sense to keep the dropout layers, they were included in the MA-SAM implementation but not mentioned in the FacT paper

Okay, I'll check out the details of it and get back on this (the easiest way would be to do experiments with both to understand this better. Let's wait on it till the first review round)

Reply:
For now, we decided to make this optional and do some tests on it's behavior with and without. Will switch the defaults later as per the experimental outcome.